### ISSUE TITLE
Modeling enumeration values semantically

### ISSUE THREAD

**Issue by fennibay** [w3c/wot-thing-description#997]
Enumerations are used commonly in automation systems, a simple example would be a property having the value range "on", "off" or "auto". Building automation in particular uses them frequently via the multistate objects in BACnet.

TD spec. defers to JSON schema for data modeling. JSON schema provides the [enum keyword](https://json-schema.org/understanding-json-schema/reference/generic.html#enumerated-values) for describing enumerations, however one can only describe primitives (numbers, strings), but cannot assign semantics to these enum values.

If we could assign semantics to enumerated values via a well-defined URI coming from an ontology, we could:
- Handle cases where different parties use different terminology, e.g. "on" vs. "active"
- Attach additional information to enumerated values such as colors, icons, translations
- map the enumerated values to the protocol requirements in protocol bindings. For instance for multistate objects on BACnet, a BACnet protocol binding could convert semantic enumerations to integers; or for a RESTful interface the semantic enumerations could be mapped to strings that this interface will understand etc.

How could this be solved? Would a simple solution such as interpreting JSON schema's enum keyword via context extension work? For instance: instead of `enum: ["on", "off", "auto"]` we would use `enum: ["myont:on", "myont:off", "myont:auto"]` where `myont` is specified in `@context`.

p.s. I use enumerated value to refer to a member (e.g. "on") of an enumeration, where enumeration refers to the whole grouping "on", "off", "auto". I think if we can model enumerated values semantically, different groupings of them in different enumerations can be sufficiently handled in the data model with existing `enum` mechanism.

# See also

Google Digital Buildings refers to this need as well: https://github.com/google/digitalbuildings/blob/master/ontology/docs/ontology.md#multi-state-values

**Comment by egekorkan** [comment]
I think this is a valid use case, however the example solution would not work since a TD consumer _should_ understand it as sending exactly `myont:on` instead of `on`. A messy solution would be:
```jsonc
{
//...
  "oneOf": [
    { "type": "string", "const":"on","@type":"myont:on" },
    { "type": "string", "const":"off","@type":"myont:off" },
    { "type": "string", "const":"auto","@type":"myont:auto" }
  ]
}
```

**Comment by fennibay** [comment]
Hi @egekorkan, many thanks for the feedback. I wasn't really hoping that that idea would work :-)

For my understanding, would the following also work?

```json
{
  "oneOf": [
    { "@id": "myont:on" },
    { "@id": "myont:off" },
    { "@id": "myont:auto" }
  ]
}
```

I tried to express here that consumer should interact using IRIs with the thing.

**Comment by egekorkan** [comment]
Theoretically yes if you want to annotate properly but `@id` will not be understood by regular JSON Schema parsers or ones who rely on the existing TD vocabulary to generate payloads or UIs. `oneOf` is already one of the least used features in the WG so I would not say that my proposal is very widely understood :/

**Comment by fennibay** [comment]
Well, `oneOf` is a standard feature of JSON Schema, so I think we can assume it will be supported. Who wouldn't understand it?

After further thinking, I understand that my essential need is enumerations of IRIs, instead of enumerations of strings or numbers. Once I have IRIs, I can relate them to an ontology, add specific protocol conversions, add translations...

So going for IRIs, this would be my attempt:

```json
{
  "oneOf": [
    { "const": {"@id": "myont:on" } },
    { "const": {"@id": "myont:off" } },
    { "const": {"@id": "myont:auto" } }
  ]
}
```

From JSON schema POV: I would require the value to be an object type of one of the three values.

Interpreting with JSON-LD support would further conclude that these are individuals with given IRIs. On this basis, I can add protocol bindings:

```json
{
    "@id": "myont:on",
    "htv:body: "on",
}
```

This way on HTTP-level simple strings (on, off, auto) would be transmitted, while a thing consumer works with the linked IRIs. I can also extend it further with translations and other mechanisms.

This is maybe a nicer solution than your "messy" one, but I understand that one would be more robust in that it would also work with non-JSON-LD-aware clients.

**Comment by egekorkan** [comment]
> Well, oneOf is a standard feature of JSON Schema, so I think we can assume it will be supported. Who wouldn't understand it?

I also agree but some are hard coding the meaning/parsing of DataSchema keywords rather than using JSON Schema based approaches.

Regarding the rest, I am not sure if I understand everything. As far as I know, `@id` is a JSON-LD feature and would not be understood by JSON Schema parsers? Also, putting `"htv:body": "on",` is bad practice in TD design since protocol related information should be only used in `forms`.

**Comment by sebastiankb** [comment]
I cannot decide whether the issue here is more or less solved or not. Otherwise we should involve members from JSON Schema or JSON-LD here.

**Comment by fennibay** [comment]
@sebastiankb, @egekorkan, sorry I couldn't respond for a while.

In summary, I think we found two alternative solutions:

### Alt. 1

Example:

```json
{
  "oneOf": [
    { "const": {"@id": "myont:on" } },
    { "const": {"@id": "myont:off" } },
    { "const": {"@id": "myont:auto" } }
  ]
}
```

1. Any kind of value, also when not part of an enumeration, should be modeled as an IRI if the author intends to have it well-defined.
1. TD, being JSON-LD, represents an IRI as `{"@id": "..."}`
1. We can construct an enum from that using the `oneOf` construct from JSON Schema
1. A non-enum could also be represented as such.
1. Further information can also be attached to this IRI, e.g.:
    1. Translations
    1. Symbols/icons
    1. Concrete representation for a protocol // should be under `forms`
    1. Also other IRIs having the same meaning can be mapped using `owl:sameAs`.
1. This doesn't cover the case where the author writes TD only as JSON and not JSON-LD. But such an author probably doesn't have requirements regarding linked data, anyway. If the TD is written as JSON-LD, yet the consumer expects only a JSON, they would see a type of object with the structure `{"@id": "..."}`. This is a general problem of using JSON-LD vs. only supporting JSON, independent from TDs, I'd say.

### Alt. 2

```json
{
  "oneOf": [
    { "const":"on","@type":"myont:on" },
    { "const":"off","@type":"myont:off" },
    { "const":"auto","@type":"myont:auto" }
  ]
}
```

// `"type": "string"` is not necessary, as the const value is a string literal

1. Use constants, and add `@type` to attach semantics.
1. This has the advantages that a JSON consumer would still understand it, and no need to map to protocol value, if they do not differ. So, backwards compatibility is better.
1. Further information can still be attached via the `@type` IRI, like alt. 1
1. I see maybe just a conceptual disadvantage, where we categorize enum values as types instead of individuals. But I don't think this is a major issue, it's just a modelling preference.

### Conclusion

We found ways to solve this problem, without needing to extend TD spec. IMHO Alt. 2 looks better.

I think we can close this issue. Many thanks for the discussion.

**Comment by egekorkan** [comment]
I also like the second alternative. However, I think that this can be included in the spec on how to provide such information on enum values. I am quite sure that there are others who would be interested. Even putting `"description"` is a requirement in the profile spec (for the core profile) so that would be a way to have description in enum values.

**Comment by fennibay** [comment]
After examining JSON-LD mechanisms further, I came up with another (IMO better) alternative:

## Alt. 3

```json
"@context": {"const": {"@type": "@vocab", "@context": {"on": "myont:On", "off": "myont:Off", "auto": "myont:Auto"}}},
"oneOf": [
    {"const": "on"},
    {"const": "off"},
    {"const": "auto"}               
]
```

1. This is similar to Alt. 1 in that the enum values are mapped to individuals and not types.
1. The syntax is simpler than Alt. 1 and 2, except for the `@context` part ;-)
1. The IRIs can be completely different than the strings. The mapping is flexible and local. I tried to show this with different casing.
1. It falls back nicely to JSON, in that we just have simple strings at the end. // Although I don't understand yet in general, how a consumer without JSON-LD support can interpret TDs.
1. `@context` could also be somewhere deeper or higher in the hierarchy.

Please also see the [example](https://json-ld.org/playground/#startTab=tab-expanded&json-ld=%7B%22%40context%22%3A%5B%22https%3A%2F%2Fwww.w3.org%2F2019%2Fwot%2Ftd%2Fv1%22%2C%7B%22myont%22%3A%22https%3A%2F%2Fexample.org%2Fmyont%23%22%7D%2C%7B%22%40version%22%3A1.1%7D%5D%2C%22id%22%3A%22urn%3Adev%3Aops%3A32473-WoTLamp-1234%22%2C%22title%22%3A%22MyLampThing%22%2C%22%40type%22%3A%22myont%3ALightSwitch%22%2C%22securityDefinitions%22%3A%7B%22basic_sc%22%3A%7B%22scheme%22%3A%22basic%22%2C%22in%22%3A%22header%22%7D%7D%2C%22security%22%3A%5B%22basic_sc%22%5D%2C%22properties%22%3A%7B%22status%22%3A%7B%22%40type%22%3A%22myont%3AOnOffState%22%2C%22%40context%22%3A%7B%22const%22%3A%7B%22%40type%22%3A%22%40vocab%22%2C%22%40context%22%3A%7B%22on%22%3A%22myont%3AOn%22%2C%22off%22%3A%22myont%3AOff%22%2C%22auto%22%3A%22myont%3AAuto%22%7D%7D%7D%2C%22oneOf%22%3A%5B%7B%22const%22%3A%22on%22%7D%2C%7B%22const%22%3A%22off%22%7D%2C%7B%22const%22%3A%22auto%22%7D%5D%2C%22forms%22%3A%5B%7B%22href%22%3A%22https%3A%2F%2Fmylamp.example.com%2Fstatus%22%7D%5D%7D%7D%7D&context=%7B%7D) in JSON-LD playground. The expanded form shows that the enum values are expanded to IRIs.

@egekorkan I understand this is all possible via context extension, foreseen in the spec. Do you think an extension to the spec is still necessary? We could provide an example, but I don't see an extension necessary at the moment.

**Comment by egekorkan** [comment]
I sadly don't understand the example but that is my problem :) However, JSON LD playground throws an error.

**Comment by fennibay** [comment]
Ok :-), I try to dissect the example:

The first `@context` makes an expanded term definition for `const` and says:
- The `@type` of values should be of `@vocab`, i.e. they shall be IRIs, but even regular texts should be mapped to the currently active vocabulary, i.e. terms in context. Without this the strings would remain as just strings, and not be converted to IRIs.
- The second, inner `@context` then adds some terms to the vocabulary, mapping the string literals on, off, auto to the desired IRIs.

Then when I use these terms as simple strings, they will be expanded to IRIs as defined by the context.

If parser doesn't understand JSON-LD, `@context` will be ignored completely, and we will have string literals.

I tried the playground again, it worked for me. What do you get as the error message?

**Comment by mcr** [comment]
This issue pointed to while discussing: https://github.com/ietf-wg-asdf/SDF/pull/8

**Comment by mjkoster** [comment]
Alternative #2 above is closest to the semantic annotation pattern we use for events, actions, properties, and data fields of complex data schemas in similar other TD examples. It also allows adding labels, descriptions, and localization hints. I'd also recommend using "anyOf" (see the referenced ASDF discussion above - ietf-wg-asdf/SDF#8 ) .
```
{
  "anyOf": [
    { 
      "description": "Manually override the automatic control and place the device in the powered state",
      "label": "on",
      "@type":"https://onedm.org/ex/sdfData/HoaType/on" 
      "const":"on"
    },
    { 
      "description": "Manually override the automatic control and place the device in the un-powered state",
      "label": "off",
      "@type":"https://onedm.org/ex/sdfData/HoaType/off" 
      "const":"off"
    },
    { 
      "description": "Apply the automatic control to the device state",
      "label": "auto",
      "@type":"https://onedm.org/ex/sdfData/HoaType/auto" 
      "const":"auto"
     }
  ]
}
```

**Comment by mjkoster** [comment]
The file uploaded to OneDM that this references looks like this:
```
{
  "info": {
    "title": "Example file for H-O-A industrial control switch", 
    "version": "2020-12-24", 
    "copyright": "Copyright 2020 Michael J. Koster. All rights reserved.", 
    "license": "https://github.com/one-data-model/oneDM/blob/master/LICENSE"
  }, 
  "namespace": {
    "ex": "https://onedm.org/exploratory/"
  }, 
  "defaultnamespace": "ex", 

  "sdfObject": {
    "HoaSwitch": {
      "sdfProperty": {
        "value": {
          "sdfRef": "ex:#/sdfData/HoaSwitchState"
        }
      }, 
      "odmAction": {
        "on": {}, 
        "off": {},
        "auto": {}
      }
    }
  },
  "sdfData": {
    "HoaSwitchState": {
      "sdfChoice": {
        "on": {
          "description": "Manually override the automatic control and place the device in the powered state",
          "label": "on",
          "default": "on"
        },
        "off": {
          "description": "Manually override the automatic control and place the device in the un-powered state",
          "label": "off",
          "default": "off"
        },
        "auto": {
          "description": "Apply the automatic control to the device state",
          "label": "auto",
          "default": "auto"
        }
      }
    }
  }
}
```

**Comment by egekorkan** [comment]
Some new discussion on this:
- `enum` is better for a developer who is working with the data schema. `oneOf` creates a more complicated structure, that is not necessary.
- We can keep `enum` but add another term that allows mapping the enum values to more semantically enriched terms. E.g. below:

```js
{
//...
"enum":[4,6,123],
"enumMap":{
  // the map uses array indexes. This way, we are not coupled to value types in enum
  "0": {"description":"Lowest speed for fan","@type":"myOnto:lowSpeed"}
  "1": {"description":"Medium speed for fan","@type":"myOnto:midSpeed"}
  "2": {"description":"Maxium speed for fan","@type":"myOnto:maxSpeed"}
  }
}
```

@lu-zero This also relates to the data mapping discussion. This is a rather common use case in bacnet devices.

**Comment by lu-zero** [comment]
The overlap between `enum` and `oneOf` is fairly annoying, for the use-case at hand using `oneOf` feels better.
I'd check with upstream jsonschema since this ambiguity should happen on a broader scope :/

in the end enum is an oneOf of consts w/out the chance to use additional metadata fields from DataSchema.

**Comment by egekorkan** [comment]
Also related to https://github.com/w3c/wot-thing-description/issues/1930. If we solve that other issue, there will be a place to add semantic information to the values.

**Comment by egekorkan** [comment]
Changing my comment from before after understanding the use case better. Here we have a simple mapping where each value in the protocol message corresponds to another value in the data schema. The bitmapping covers the more complex case.

**Comment by fennibay** [comment]
@egekorkan Yes, bitmapping a little bit more complex, but IMHO still addressable with enumeration mapping. 🙂 

The essential problem is assigning additional attributes to some numbers or strings (semantic meaning/IRI, color, icon etc.):

```json
    "hasEnumMap": {
        "hasMapMember": {
            "1": { "@id": "example:on" },
            "2": { "@id": "example:off" },
            "3": { "@id": "example:auto" }
        }
    }
```

Once that's in place, an enum property is simply modeled as:

```json
    "enum": [1, 2, 3]
```

And a bitmap property can be simply modeled as:

```json
    "type": "array",
    "items": {"enum": [1, 2, 3]},
    "uniqueItems": true
```

This is a bit counterintuitive at first look, as some 32/64-bit bitstring is transferred over the wire, and people try to map it to the InteractionAffordance level. But I find this is the most agnostic representation, it avoids a lot of problematic details like endianness, can represent sparse bitstrings very well, and is not limited in length. The only downside is that the protocol-representation should convert the protocol-specific encoding to this array of numbers.
